Fix compilation warnings

fwyzard opened the pull request:
Make sure `pragma clang diagnostic` is only used when compiling with clang, and not with gcc.

With HIP version 5.2.0 or newer, include `hiprand/hiprand_kernel.h` instead of `hiprand_kernel.h`.

fwyzard committed:
Fix compilation warnings

Make sure "pragma clang diagnostic" is only used when compiling with
clang, and not with gcc.

With HIP version 5.2.0 or newer, include "hiprand/hiprand_kernel.h"
instead of "hiprand_kernel.h".

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

j-stephan commented:
Hi @fwyzard, thanks for the submission. Please note that the AMD CI will likely fail at the moment - we had a power surge which unfortunately grilled at least one of our AMD GPUs.

fwyzard commented:
Ops, sorry to hear that :-(

fwyzard committed:
Replace BOOST_LANG_HIP with HIP_VERSION

BOOST_LANG_HIP is available only when compiling HIP code, while
HIP_VERSION is available also while using the HIP host API.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard commented:
The last change is similar to what I did for CUDA some time ago, and is needed to get a consistent behaviour between `hipcc` (compiling device or host code) and `gcc` (compiling host-only code that calls the HIP API, like allocating GPU memory).

psychocoderHPC commented on include/alpaka/core/ApiHipRt.hpp:
```diff
@@ -11,17 +11,17 @@
         // Names
         static constexpr char name[] = "Hip";
-        static constexpr auto version = BOOST_LANG_HIP;
+        static constexpr auto version = BOOST_VERSION_NUMBER(HIP_VERSION_MAJOR, HIP_VERSION_MINOR, 0);
```
Why do you switch from BOOST_VERSION to the version format ROCM is shipping? Is this necessary?

I am asking because normally we use always BOOST_VERSIONS in alpaka and now we need to take care to use for HIP the HIP internal format.

fwyzard replied:
hi @psychocoderHPC, I'm not sure I follow, are you asking why I used the BOOST_VERSION macro explicitly, or why I didn't include the third field in it ?

psychocoderHPC replied:
I misread this line I thought you do not use `BOOST_VERSION` anymore but you use it. I was confused by the usage later, I will mark the line.

psychocoderHPC commented on include/alpaka/core/ApiHipRt.hpp:
```diff
@@ -83,7 +83,7 @@ namespace alpaka
         static constexpr DeviceAttr_t deviceAttributeMultiprocessorCount = ::hipDeviceAttributeMultiprocessorCount;
 
-#    if BOOST_LANG_HIP >= BOOST_VERSION_NUMBER(4, 5, 0)
+#    if HIP_VERSION >= 40500000
```
Regarding to my last point, I am confused why you use `HIP_VERSION` and not `BOOST_LANG_VERSION` as before.
What was wrong with the original macro comparison?

fwyzard replied:
The problem is that `BOOST_LANG_HIP` is defined only when using `hipcc` to compile host or device code, but it does not get defined when using a different compiler (_e.g._ `gcc`) to compile HIP host code.

The same can be done using the Boost version macros:
```suggestion
#    if BOOST_VERSION_NUMBER(HIP_VERSION_MAJOR, HIP_VERSION_MINOR, 0) >= BOOST_VERSION_NUMBER(4, 5, 0)
```

I can change it to this approach if it is preferred (and then I guess I should make the same change in `include/alpaka/core/ApiCudaRt.hpp` with respect to `CUDART_VERSION`).

fwyzard replied:
Using `HIP_VERSION` is more compact :-)

psychocoderHPC replied:
IMO we should keep `if BOOST_LANG_HIP >= BOOST_VERSION_NUMBER(4, 5, 0)` to have a single place in the code where the version number stored, in our case it is `#include <alpaka/core/BoostPredef.hpp>`

psychocoderHPC replied:
We discussed this topic in the dev meeting and we should unify the creation of the version number in `BoostPredef.hpp` so that we do not have `BOOST_VERSION_NUMBER(HIP_VERSION_MAJOR, HIP_VERSION_MINOR, 0)` somewhere in the code expect in the predef header.

`BOOST_LANG_HIP` should be set always if `HIP_VERSION_MAJOR` is available.

fwyzard replied:
This is the only solution that _does not work_ .

`BOOST_LANG_HIP` is defined only when compiling with `hipcc`, not when compiling with `gcc`, so it is not a suitable macro to use here.

`BOOST_LANG_HIP`'s behaviour cannot be changed, as it is used elsewhere specifically to check for a compiler with HIP support.

The viable options are
  - use `HIP_VERSION >= 40500000`
  - use `BOOST_VERSION_NUMBER(HIP_VERSION_MAJOR, HIP_VERSION_MINOR, 0) >= BOOST_VERSION_NUMBER(4, 5, 0)`
  - `#define BOOST_HIP_VERSION = BOOST_VERSION_NUMBER(HIP_VERSION_MAJOR, HIP_VERSION_MINOR, 0)` and check `BOOST_HIP_VERSION >= BOOST_VERSION_NUMBER(4, 5, 0)`

fwyzard replied:
And, whatever the choice, it should be applied consistently here and in `include/alpaka/core/ApiCudaRt.hpp` (with respect to `BOOST_LANG_CUDA` vs `CUDART_VERSION`).

fwyzard replied:
> `BOOST_LANG_HIP` should be set always if `HIP_VERSION_MAJOR` is available.

Maybe - but that is a _huge_ change in how Alpaka headers are written, and a departure from how Boost's own `BOOST_LANG_CUDA` behaves: `BOOST_LANG_CUDA` is defined only when compiling a `.cu` file with `nvcc` (or `clang` in cuda mode).

psychocoderHPC replied:
OK, in that case, keep it how it is for now and we will postpone it to a future PR.
